### ISSUE TITLE
Script resolution depends on order of element in config file

### ISSUE THREAD

**Issue by ppkarwasz** [apache/logging-log4j2#3336]
The resolution of `ScriptRef` elements depends on the order of elements in the configuration file: the `Scripts` element must appear **before** any `ScriptRef` element.

**Comment by Suvrat1629** [comment]
(I am new to open source contribution and trying to understand the code)
Are the changes mentioned in the issue to be made in the log4j2.xml file or the log4j2.json?

**Comment by ppkarwasz** [comment]
The fix must be done in code.

The basic problem is that the resolution of script references is done in constructors or factory methods of script-enabled components. For example in [`ScriptFilter`](https://logging.apache.org/log4j/2.x/manual/filters.html#Script) we have this snippet in the `createFilter` method:

https://github.com/apache/logging-log4j2/blob/5496c7250184e6460651eb966e8913dd0f3fe341/log4j-core/src/main/java/org/apache/logging/log4j/core/filter/ScriptFilter.java#L148-L157

Since the constructors of Log4j components are called in the order the components are declared in the configuration file, something like this will fail:

```xml
<ScriptFilter>
  <ScriptRef ref="GLOBAL_FILTER"/>
</ScriptFilter>
<Scripts>
  <ScriptFile name="GLOBAL_FILTER"
              language="groovy"
              path="global-filter.groovy"/>
</Scripts>
```

will fail, since at the moment `ScriptRef` is resolved, the `ScriptFile` element has not been processed yet. To solve this problem, the resolution of `ScriptRef`s must be delayed. For example you can perform the resolution in the `start()` method of the [`LifeCycle`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/LifeCycle.html) interface.

**Comment by Suvrat1629** [comment]
Ok, I did make the changes as you suggested by using the start() method. What more changes do i need to make now? Do i need to change the testcases or anything?

**Comment by ppkarwasz** [comment]
@Suvrat1629,

Yes, if you want to contribute the fix, you should add some tests. See the [PR template](https://github.com/apache/logging-log4j2/blob/2.x/.github/pull_request_template.md) for more details.

Should I assign the issue to you?

**Comment by Suvrat1629** [comment]
Yes you can assign it to me, also can you guide me with the test cases changes as the ScriptFilterTest.java seems to be consistent with my code is there any file that i need to look into?

**Comment by vy** [comment]
Fixed by #3718.
